tokio/src/macros/select.rs

```diff
@@ -131,6 +131,10 @@
 /// correctly even if it is restarted while waiting at an `.await`, then it is
 /// cancellation safe.
 ///
+/// Cancellation safety can be defined in the following way: If you have a
+/// future that has not yet completed, then it must be a no-op to drop that
+/// future and recreate it.
```

Comment:
If this definition ought to be very precise, I would probably refrain from defining it through the observable effects of future drop + recreate. For example, I would say the following future is cancellation safe under this definition, but just dropping it halfway through the download without driving to the completion can leak disk space:
```rust
async fn download_and_verify() -> Sha256Sum {
    for i in 0..N {
        if !is_chunk_downloaded(i) {
            download_data_chunk(i).await;
        }
    }
    let sha_sum = compute_hash();
    delete_all_data_chunks();
    sha_sum
}
```
In my opinion, it is not safe to use this future inside the select! macro, please let me know if you agree.

Reply:
How would you define it? Inserting a drop-and-recreate of your future in the middle of a program _does_ change what it does, so in some sense it is not a no-op, even if the difference is not observable until later.

Reply:
True, I agree that it may not be considered as a no-op.

I was going to say that I would define it like this: “a future is cancellation safe if and only if creating it and dropping before completion is a no-op”. Now that I think of it, it does not account for losing a place in a fair queue, for example in `Mutex::lock`.

Maybe the definition should be more rigorous and state that both create-and-drop *and* drop-and-recreate should be no-ops? (with “no-op” meaning “no observable effects”)

Reply:
I expanded a bit, but I think it's fine as-is.